src/components/implementation/no_interface/scheddev/sl.h

```diff
@@ -174,11 +175,177 @@ sl_cs_exit(void)
+ * timeout relative to now, and the oneshot timeout sets a timeout on
+ * the same absolute timeline as returned by sl_now.
+ */
+void sl_timeout_period(cycles_t period);
+
+static inline cycles_t
+sl_timeout_period_get(void)
+{ return sl__globals()->period; }
+
+static inline void
+sl_timeout_oneshot(cycles_t absolute_us)
+{
+	sl__globals()->timer_next   = absolute_us;
+	sl__globals()->timeout_next = tcap_cyc2time(absolute_us);
+}
+
+static inline void
+sl_timeout_relative(cycles_t offset)
+{ sl_timeout_oneshot(sl_now() + offset); }
+
+static inline microsec_t
+sl_cyc2usec(cycles_t cyc)
+{ return cyc / sl__globals()->cyc_per_usec; }
+
+static inline microsec_t
+sl_usec2cyc(microsec_t usec)
+{ return usec * sl__globals()->cyc_per_usec; }
+
+/* Timeout and wakeup functionality */
+#define SL_TIMEOUT_HEAP()       (&timeout_heap.h)
+
+struct timeout_heap {
+	struct heap  h;
+	void        *data[SL_MAX_NUM_THDS];
+} timeout_heap;
+
+/* wakeup any blocked threads! */
+static void
+sl_timeout_wakeup_expired(cycles_t now)
+{
+	if (!heap_size(SL_TIMEOUT_HEAP())) return;
+
+	do {
+		struct sl_thd *tp, *th;
+
+		tp = heap_peek(SL_TIMEOUT_HEAP());
+		assert(tp);
+
+		/* FIXME: logic for wraparound in current tsc */
+		if (tp->timeout_cycs > now) break;
+
+		th = heap_highest(SL_TIMEOUT_HEAP());
+		assert(th && th == tp);
+		th->timeout_idx = -1;
+
+		assert(th->wakeup_cycs == 0);
+		th->wakeup_cycs = now;
+		sl_thd_wakeup_no_cs(th);
+	} while (heap_size(SL_TIMEOUT_HEAP()));
+}
+
+static inline void
+sl_timeout_block(struct sl_thd *t, cycles_t timeout)
+{
+	assert(t && t->timeout_idx == -1); /* not already in heap */
+	assert(heap_size(SL_TIMEOUT_HEAP()) < SL_MAX_NUM_THDS);
+
+	if (!timeout) {
+		cycles_t tmp = t->periodic_cycs;
+
+		assert(t->period);
+		t->periodic_cycs += t->period; /* implicit timeout = task period */
+		assert(tmp < t->periodic_cycs); /* wraparound check */
+		t->timeout_cycs   = t->periodic_cycs;
+	} else {
+		t->timeout_cycs   = timeout;
+	}
+
+	t->wakeup_cycs = 0;
+	heap_add(SL_TIMEOUT_HEAP(), t);
+}
+
+static inline void
+sl_timeout_remove(struct sl_thd *t)
+{
+	assert(t && t->timeout_idx > 0);
+	assert(heap_size(SL_TIMEOUT_HEAP())); 
+
+	heap_remove(SL_TIMEOUT_HEAP(), t->timeout_idx);
+	t->timeout_idx = -1;
+}
+
+static inline void
+sl_timeout_expended(microsec_t now, microsec_t oldtimeout)
+{
+	cycles_t offset;
+
+	assert(now >= oldtimeout);
+
+	/* in virtual environments, or with very small periods, we might miss more than one period */
+	offset = (now - oldtimeout) % sl_timeout_period_get();
+	sl_timeout_oneshot(now + sl_timeout_period_get() - offset);
+}
+
+static inline int
+__sl_timeout_compare_min(void *a, void *b)
+{
+	/* FIXME: logic for wraparound in either timeout_cycs */
+	return ((struct sl_thd *)a)->timeout_cycs <= ((struct sl_thd *)b)->timeout_cycs;
+}
+
+static inline void
+__sl_timeout_update_idx(void *e, int pos)
+{ ((struct sl_thd *)e)->timeout_idx = pos; }
+
+static inline void
```

Comment:
I understand what you mean here: `init` is not called often and it's not on the `fast-path`, so I should move it to `.c` and omit `inline`. Perhaps GCC will be smart enough to inline as part of it's optimizations even otherwise. 

Will change it. Thank you.